src/rpc/AppRpcProvider.test.ts

```diff
@@ -0,0 +1,108 @@
+import { TransactionResponse } from '@ethersproject/abstract-provider'
+import { JsonRpcProvider, Network, Provider } from '@ethersproject/providers'
+
+import AppRpcProvider from './AppRpcProvider'
+
+jest.mock('@ethersproject/providers')
+
+describe('AppRpcProvider', () => {
+  let mockProviders: JsonRpcProvider[]
+  let mockProvider1: jest.Mocked<JsonRpcProvider>
+  let mockProvider2: jest.Mocked<JsonRpcProvider>
+  let mockIsProvider: jest.SpyInstance
+
+  beforeEach(() => {
+    mockProvider1 = new JsonRpcProvider() as jest.Mocked<JsonRpcProvider>
+    mockProvider2 = new JsonRpcProvider() as jest.Mocked<JsonRpcProvider>
+
+    mockIsProvider = jest.spyOn(Provider, 'isProvider').mockReturnValue(true)
+    mockProviders = [mockProvider1, mockProvider2]
+    mockProviders.forEach((provider) => {
+      // override readonly property
+      // @ts-expect-error
+      provider.network = {
+        name: 'homestead',
+        chainId: 1,
+      } as Network
+      provider.getNetwork = jest.fn().mockReturnValue({
+        name: 'homestead',
+        chainId: 1,
+      } as Network)
+    })
+  })
+
+  test('constructor initializes with valid providers', () => {
+    expect(() => new AppRpcProvider(mockProviders)).not.toThrow()
+  })
+
+  test('constructor throws with empty providers array', () => {
+    expect(() => new AppRpcProvider([])).toThrow('providers array empty')
+  })
+
+  test('constructor throws with network mismatch', () => {
+    mockProviders[0].network.chainId = 2
+    expect(() => new AppRpcProvider(mockProviders)).toThrow('networks mismatch')
+  })
+
+  test('constructor throws with invalid providers', () => {
+    mockIsProvider.mockReturnValueOnce(false)
+    expect(() => new AppRpcProvider([{} as JsonRpcProvider])).toThrow('invalid provider')
+  })
+
+  test('handles sendTransaction', async () => {
+    const hash = '0x123'
+    mockProvider1.sendTransaction.mockResolvedValue({ hash } as TransactionResponse)
+    const provider = new AppRpcProvider([mockProvider1])
+
+    const result = await provider.perform('sendTransaction', { signedTransaction: '0xabc' })
+    expect(result).toBe(hash)
+  })
+
```

Comment:
Add a test for sends which are not sendTransaction.
Add a test for sends which fail, as they currently will not fall back, because they are not caught with an `await`.